components/codec/src/columnar/bool.rs

```diff
@@ -0,0 +1,245 @@
+///   is enabled.
+struct Encoding;
+
+/// The compression for [`Encoding`].
+///
+/// It is not allowed to be modified and only allowed to be appended with a new
+/// variant.
+#[derive(Clone, Copy, Default)]
+#[repr(C)]
+enum Compression {
+    #[default]
+    None = 0,
+    BitSet = 1,
+}
+
+impl Encoding {
+    const COMPRESSION_SIZE: usize = 1;
+    /// The overhead for compression is 5B, so it is not good to always enable
+    /// the compression.
+    const COMPRESS_THRESHOLD: usize = 10;
+    const NUM_VALUES_SIZE: usize = 4;
+    const VERSION: u8 = 0;
+    const VERSION_SIZE: usize = 1;
+
+    fn need_compress(num_values: usize) -> bool {
+        num_values > Self::COMPRESS_THRESHOLD
+    }
+
+    fn decode_compression(flag: u8) -> Result<Compression> {
+        let compression = match flag {
+            0 => Compression::None,
+            1 => Compression::BitSet,
+            _ => InvalidCompression { flag }.fail()?,
+        };
+
+        Ok(compression)
+    }
+
+    fn encode<B, I>(&self, buf: &mut B, values: I) -> Result<()>
+    where
+        B: BufMut,
+        I: Iterator<Item = bool> + Clone,
+    {
+        buf.put_u8(Self::VERSION);
+
+        let num_values = values.clone().count();
+        if Self::need_compress(num_values) {
+            Self::encode_with_compression(buf, num_values, values)
+        } else {
+            Self::encode_without_compression(buf, values)
+        }
+    }
+
+    fn estimated_encoded_size<I>(&self, values: I) -> usize
+    where
+        I: Iterator<Item = bool>,
+    {
+        let num_values = values.count();
+        if Self::need_compress(num_values) {
+            BitSet::num_bytes(num_values)
+                + Self::COMPRESSION_SIZE
+                + Self::NUM_VALUES_SIZE
+                + Self::VERSION_SIZE
+        } else {
+            num_values + Self::VERSION_SIZE + Self::COMPRESSION_SIZE
+        }
+    }
+
+    fn decode<B, F>(&self, buf: &mut B, f: F) -> Result<()>
+    where
+        B: Buf,
+        F: FnMut(bool) -> Result<()>,
+    {
+        let buf = buf.chunk();
+        ensure!(
+            buf.len() > Self::VERSION_SIZE + Self::COMPRESSION_SIZE,
+            NotEnoughBytes { len: buf.len() }
+        );
+
+        // Decode the version.
+        let version = buf[0];
+        ensure!(version == Self::VERSION, InvalidVersion { version });
+
+        // Decode the compression.
+        let compression_index = buf.len() - 1;
+        match Self::decode_compression(buf[compression_index])? {
+            Compression::None => Self::decode_without_compression(buf, f)?,
+            Compression::BitSet => Self::decode_with_compression(buf, f)?,
+        }
+
+        Ok(())
+    }
+
+    fn encode_without_compression<B, I>(buf: &mut B, values: I) -> Result<()>
+    where
+        B: BufMut,
+        I: Iterator<Item = bool>,
+    {
+        for v in values {
+            buf.put_u8(v as u8);
```

Comment:
use bit to represent bool?

Reply:
This encoding method is chosen when the number of boolean is small, which may lead to
low compression ratio.